core/primitives/src/shard_layout.rs

```diff
@@ -83,17 +96,195 @@ pub struct ShardLayoutV1 {
+        let range = format!("{:?}..{:?}", self.start, self.end);
+        f.debug_struct("AccountRange").field("range", &range).finish()
+    }
+}
+
+impl AccountRange {
+    pub fn new_single_shard() -> Self {
+        Self { start: AccountBoundary::Start, end: AccountBoundary::End }
+    }
+
+    pub fn new_start(end: &str) -> Self {
+        Self { start: AccountBoundary::Start, end: AccountBoundary::Middle(end.parse().unwrap()) }
+    }
+
+    pub fn new_end(start: &str) -> Self {
+        Self { start: AccountBoundary::Middle(start.parse().unwrap()), end: AccountBoundary::End }
+    }
+
+    pub fn new_mid(start: &str, end: &str) -> Self {
+        Self {
+            start: AccountBoundary::Middle(start.parse().unwrap()),
+            end: AccountBoundary::Middle(end.parse().unwrap()),
+        }
+    }
+
+    fn contains(&self, account_id: &AccountId) -> bool {
+        match (&self.start, &self.end) {
+            (AccountBoundary::Start, AccountBoundary::End) => true,
+            (AccountBoundary::Start, AccountBoundary::Middle(end)) => account_id < end,
+            (AccountBoundary::Middle(start), AccountBoundary::End) => account_id >= start,
+            (AccountBoundary::Middle(start), AccountBoundary::Middle(end)) => {
+                account_id >= start && account_id < end
+            }
+            (AccountBoundary::End, _) => panic!("invalid account range"),
+            (_, AccountBoundary::Start) => panic!("invalid account range"),
+        }
+    }
+}
+
+/// A mapping from the shard id to the account range of this shard.
+type ShardsAccountRange = BTreeMap<ShardId, AccountRange>;
+
+fn validate_shards_account_range(
+    shards_account_range: &ShardsAccountRange,
+) -> Result<(), ShardLayoutError> {
+    let err =
+        |reason: &str| ShardLayoutError::InvalidShardsAccountRange { reason: reason.to_string() };
+
+    let values = shards_account_range.values().sorted().collect_vec();
+    println!("{:?}", values);
+
+    let (Some(first), Some(last)) = (values.first(), values.last()) else {
+        return Err(err("account range empty"));
+    };
+    if first.start != AccountBoundary::Start {
+        return Err(err("first account range should start with Start"));
+    }
+    if last.end != AccountBoundary::End {
+        return Err(err("last account range should end with End"));
+    }
+
+    for i in 1..values.len() {
+        let prev = values[i - 1];
+        let curr = values[i];
+        if prev.end != curr.start {
+            return Err(err("account ranges should be contiguous"));
+        }
+    }
+
+    Ok(())
+}
+
+/// Making the shard ids non-contiguous.
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, PartialEq, Eq)]
+pub struct ShardLayoutV2 {
+    /// A mapping from the shard id to the account range of this shard.
+    shards_account_range: ShardsAccountRange,
+
+    /// A mapping from the parent shard to child shards. Maps shards from the
```

Comment:
I'm wondering if it makes sense to have this struct contain all historic mappings instead of just the latest split?

That would be quite useful for store layer where we may recursively need to find parent shard_ids

Reply:
Maybe but let's keep it simple for now and we can enhance it later as needed.